Sources/Logging/Logging.swift

```diff
@@ -653,55 +661,165 @@ extension Logger {
+    internal static var metadataProvider: Logger.MetadataProvider {
+        return self._metadataProviderFactory.metadataProvider
     }
 
     private final class FactoryBox {
         private let lock = ReadWriteLock()
-        fileprivate var _underlying: (String) -> LogHandler
+        fileprivate var _underlying: (String, Logger.MetadataProvider) -> LogHandler
         private var initialized = false
 
-        init(_ underlying: @escaping (String) -> LogHandler) {
+        init(_ underlying: @escaping (String, Logger.MetadataProvider) -> LogHandler) {
             self._underlying = underlying
         }
 
-        func replaceFactory(_ factory: @escaping (String) -> LogHandler, validate: Bool) {
+        func replaceFactory(_ factory: @escaping (String, Logger.MetadataProvider) -> LogHandler, validate: Bool) {
             self.lock.withWriterLock {
                 precondition(!validate || !self.initialized, "logging system can only be initialized once per process.")
                 self._underlying = factory
                 self.initialized = true
             }
         }
 
-        var underlying: (String) -> LogHandler {
+        var underlying: (String, Logger.MetadataProvider) -> LogHandler {
+            return self.lock.withReaderLock {
+                return self._underlying
+            }
+        }
+    }
+
+    private final class MetadataProviderBox {
+        private let lock = ReadWriteLock()
+
+        internal var _underlying: Logger.MetadataProvider
+        private var initialized = false
+
+        init(_ underlying: Logger.MetadataProvider) {
+            self._underlying = underlying
+        }
+
+        func replaceMetadataProvider(_ metadataProvider: Logger.MetadataProvider, validate: Bool) {
+            self.lock.withWriterLock {
+                precondition(!validate || !self.initialized, "logging system can only be initialized once per process.")
+                self._underlying = metadataProvider
+                self.initialized = true
+            }
+        }
+
+        var metadataProvider: Logger.MetadataProvider {
             return self.lock.withReaderLock {
                 return self._underlying
             }
         }
     }
 }
 
-extension Logger {
+public extension LoggingSystem {
+    /// `bootstrap` is a one-time configuration function which globally selects the desired logging backend
+    /// implementation. `bootstrap` can be called at maximum once in any given program, calling it more than once will
+    /// lead to undefined behavior, most likely a crash.
+    ///
+    /// - parameters:
+    ///     - metadataProvider: The `MetadataProvider` used to inject runtime-generated metadata, defaults to a "no-op" provider.
+    ///     - factory: A closure that given a `Logger` identifier, produces an instance of the `LogHandler`.
+    static func bootstrap(metadataProvider: Logger.MetadataProvider,
```

Comment:
feels like the metadataProvider should be the second argument

Reply:
A while ago we came to the order of parameters recommendations around "first strict, then closures, then optional params, then optional closures" so this follows that recommendation.

I'm not sure how many people actually write 

```
LoggingSystem.bootstrap {
  many
  lines
  return here
}
```

so it was intended to keep that shape 🤔 

```
LoggingSystem.bootstrap(metadataProvider: .coolTracer) {
  many
  lines
  return here
}
```

if we flip it, we end up with:

```
LoggingSystem.bootstrap({
  many
  lines
  return here
}, metadataProvider: .coolTracer)
```

potentially... but arguably most uses do look like this:

```
LoggingSystem.bootstrap(.coolLogger, metadataProvider: .coolTracer)
```

potentially... so, I'm not really sure about this one. The current shape kind of follows the "usual" order, with closures last. Open to either but current seemed more natural given the usage shapes hm

Reply:
Okey, let's do it 😄